<a href="https://colab.research.google.com/github/rdkrishna/ML_Docs_Portfolio/blob/main/MultiVariate_Classification_Hackathon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Introduction and Details of Features of data**
This is Hackathon project to predict (Classify) 2 outputs,
Primary and Final close flags, binary indicators representing default or non default (representing 1 and 0 respectively) of credict at end of primary term and final term.
Hackthon details is Analytics Olympiad 2023 by MachineHack and Shiv Nadar Institute of Emminence.
The data set is available in link: https://machinehack.com/hackathons/analytics_olympiad_2023/data

The Data is already split as Train and Test datasets.

So, in this scenario, we split the Train data set as X and y, to train, test and validate internally and best model is then validated against Test datasets

## Features and Labels:
days_since_opened: Days from credit opening date to data collection date.
days_since_confirmed: Days from credit information confirmation date till data collection date.
primary_term: Planned number of days from credit opening date to closing date.
final_term: Actual number of days from credit opening date to closing date.
days_till_primary_close: Planned number of days from data collection date until loan closing date.
days_till_final_close: Actual number of days from data collection date until loan closing date.
loans_credit_limit: Credit limit for the customer's loans.
loans_next_payment_summary: Amount of the next loan payment.
loans_outstanding_balance: Outstanding balance amount.
loans_max_overdue_amount: Maximum overdue amount.
loans_credit_cost_rate: Cost rate associated with loans.

Loan Overdue Information:

loans_within_5_days to loans_over_90_days: Number of loans overdue within different time frames.
is_zero_loans_within_5_days to is_zero_loans_over_90_days: Binary indicators for zero loans overdue within different time frames.

Credit Utilization and Limit Information:

utilization: Credit utilization rate.
over_limit_count: Count of instances where the customer went over the credit limit.
max_over_limit_count: Maximum count of instances where the customer exceeded the credit limit.
is_zero_utilization: Binary indicator if credit utilization rate is zero.
is_zero_over_limit_count: Binary indicator if the count of over limit instances is zero.
is_zero_max_over_limit_count: Binary indicator if the maximum over limit count is zero.

Encoded Features:

encoded_payment_X: Encoded information about payment X (categorical features converted to numerical values).
encoded_loans_account_holder_type: Encoded information about the type of account holder for loans.
encoded_loans_credit_status: Encoded information about the credit status of loans.
encoded_loans_credit_type: Encoded information about the type of credit for loans.
encoded_loans_account_currency: Encoded information about the currency used for loans.

Close Flags: (Labels)

primary_close_flag: Binary indicator for primary term close.
final_close_flag: Binary indicator for final term close.

# Importing Libraries and Datasets

In [1]:
import pandas as pd
import io
import numpy as np
from matplotlib import pyplot as plt
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import LabelEncoder

In [2]:
# For Model building and Training
from sklearn.metrics import accuracy_score, classification_report,f1_score
from sklearn.metrics import multilabel_confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.model_selection import RandomizedSearchCV
from sklearn.neighbors import KNeighborsClassifier


In [3]:
import time
import warnings
warnings.filterwarnings("ignore")

In [4]:
# GPU Execution Prerequsite
import sys
import os
import site
sys.path.append(os.path.join(os.path.dirname(site.getsitepackages()[0]), "site-packages"))


In [ ]:
# !wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
# !chmod +x Miniconda3-latest-Linux-x86_64.sh
# !bash ./Miniconda3-latest-Linux-x86_64.sh -b -f -p /usr/local
# import sys
# sys.path.append('/usr/local/lib/python3.7/site-packages/')
# ! conda install -c rapidsai -c nvidia -c conda-forge \
#     -c defaults rapids=0.14 python=3.7 cudatoolkit=10.1


/bin/bash: line 1: conda: command not found


In [5]:
# # For Plotting
from sklearn.metrics import roc_curve, auc
import matplotlib.pyplot as plt

In [6]:
# Plot and Other Display Settings.
pd.options.display.max_rows = 10
pd.options.display.float_format = "{:.1f}".format

In [8]:
# from google.colab import drive
# drive.mount('/content/drive')

In [44]:
training_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/train.csv')

In [ ]:
training_df.head()

customer_id firstname lastname  record_number  \
0  4814dfa2-45a8-46b9-8102-98ecbbec2d2a     Navya     Vasa              7   
1  bb0abe41-cc89-4a1d-bb1c-48bd212ae00d      Azad  Warrior              5   
2  edfd1a62-05da-4cfd-b8ae-521f05f0fd5f      Piya      Jha              1   
3  3eaa4c2a-2316-48f4-b1b0-4aafe2286630      Rhea    Varma             11   
4  c9a73239-baae-40eb-8855-dab07767ea86     Sahil     Dave             12   

   days_since_opened  days_since_confirmed  primary_term  final_term  \
0                 10                     8            12          10   
1                 15                     6             7           6   
2                 15                    17             8           5   
3                  7                     2            11          13   
4                 14                     9             8          11   

   days_till_primary_close  days_till_final_close  ...  encoded_payment_21  \
0                       13                      5  ...                 0.0   
1                        9                      1  ...                 3.0   
2                        9                      1  ...                 3.0   
3                       14                      8  ...                 3.0   
4                        6                     13  ...                 3.0   

   encoded_payment_22  encoded_payment_23  encoded_payment_24  \
0                 1.0                 3.0                 4.0   
1                 3.0                 3.0                 4.0   
2                 3.0                 3.0                 4.0   
3                 3.0                 3.0                 4.0   
4                 3.0                 3.0                 4.0   

   encoded_loans_account_holder_type  encoded_loans_credit_status  \
0                                  1                            3   
1                                  1                            3   
2                                  1                            3   
3                                  1                            2   
4                                  1                            3   

   encoded_loans_credit_type  encoded_loans_account_currency  \
0                          4                               1   
1                          4                               1   
2                          4                               1   
3                          4                               1   
4                          4                               1   

   primary_close_flag  final_close_flag  
0                   0                 0  
1                   0                 0  
2                   0                 0  
3                   0                 0  
4                   0                 0  

[5 rows x 62 columns]

In [11]:
testing_df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/test.csv')

In [45]:
training_df.shape

(1106674, 62)

In [12]:
testing_df.shape

(474289, 60)

training_df.shape = (1106674, 62)
> 1.1 Million Records

testing_df.shape = (474289, 60)
> 0.47 Million Records



Now, Loaded the data into dataframe, Step2:

This is a classification Problem.
2 Labels - primary_close_flag
         - final_close_flag

In [ ]:
training_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1106674 entries, 0 to 1106673
Data columns (total 62 columns):
 #   Column                              Non-Null Count    Dtype  
---  ------                              --------------    -----  
 0   customer_id                         1106674 non-null  object 
 1   firstname                           1106674 non-null  object 
 2   lastname                            1106674 non-null  object 
 3   record_number                       1106674 non-null  int64  
 4   days_since_opened                   1106674 non-null  int64  
 5   days_since_confirmed                1106674 non-null  int64  
 6   primary_term                        1106674 non-null  int64  
 7   final_term                          1106674 non-null  int64  
 8   days_till_primary_close             1106674 non-null  int64  
 9   days_till_final_close               1106674 non-null  int64  
 10  loans_credit_limit                  1106674 non-null  int64  
 11  loans_next_

# Data Preprocessing

1. Shows that Data is missing in some cells of column encoded_payment_n
Defining function and filling the value.

Creating generic function which can be used for different learning Models, by passing model as parameter.

### Step 1: Defining Preprocessing Functions

In [ ]:
dict(training_df.isnull().sum())

{'customer_id': 0,
 'firstname': 0,
 'lastname': 0,
 'record_number': 0,
 'days_since_opened': 0,
 'days_since_confirmed': 0,
 'primary_term': 0,
 'final_term': 0,
 'days_till_primary_close': 0,
 'days_till_final_close': 0,
 'loans_credit_limit': 0,
 'loans_next_payment_summary': 0,
 'loans_outstanding_balance': 0,
 'loans_max_overdue_amount': 0,
 'loans_credit_cost_rate': 0,
 'loans_within_5_days': 0,
 'loans_within_5_to_30_days': 0,
 'loans_within_30_to_60_days': 0,
 'loans_within_60_to_90_days': 0,
 'loans_over_90_days': 0,
 'is_zero_loans_within_5_days': 0,
 'is_zero_loans_within_5_to_30_days': 0,
 'is_zero_loans_within_30_to_60_days': 0,
 'is_zero_loans_within_60_to_90_days': 0,
 'is_zero_loans_over_90_days': 0,
 'utilization': 0,
 'over_limit_count': 0,
 'max_over_limit_count': 0,
 'is_zero_utilization': 0,
 'is_zero_over_limit_count': 0,
 'is_zero_max_over_limit_count': 0,
 'encoded_payment_0': 37619,
 'encoded_payment_1': 35744,
 'encoded_payment_2': 34612,
 'encoded_payment_3'

In [13]:
def fill_missing_value(df):
    # columns_with_missing_values = ['encoded_payment_0','encoded_payment_1','encoded_payment_2',
    #                                'encoded_payment_3','encoded_payment_4','encoded_payment_5',
    #                                'encoded_payment_6','encoded_payment_7','encoded_payment_8',
    #                                'encoded_payment_9','encoded_payment_10','encoded_payment_11',
    #                                'encoded_payment_12','encoded_payment_13','encoded_payment_14',
    #                                'encoded_payment_15','encoded_payment_16','encoded_payment_17',
    #                                'encoded_payment_18','encoded_payment_19','encoded_payment_20',
    #                                'encoded_payment_21','encoded_payment_22',
    #                                'encoded_payment_23','encoded_payment_24']
    encoded_features = ['encoded_payment_' + str(i) for i in range(25)]
    imputer = SimpleImputer(strategy ='most_frequent')
    # df[columns_with_missing_values] = imputer.fit_transform(df[columns_with_missing_values])
    df[encoded_features] = imputer.fit_transform(df[encoded_features])

    return df

####Commenting Unwanted/test Codes

In [ ]:
# training_df = fill_missing_value(training_df)

In [ ]:
# columns_with_missing_values = ['encoded_payment_0','encoded_payment_1','encoded_payment_2',
#                                    'encoded_payment_3','encoded_payment_4','encoded_payment_5',
#                                    'encoded_payment_6','encoded_payment_7','encoded_payment_8',
#                                    'encoded_payment_9','encoded_payment_10','encoded_payment_11',
#                                    'encoded_payment_12','encoded_payment_13','encoded_payment_14',
#                                    'encoded_payment_15','encoded_payment_16','encoded_payment_17',
#                                    'encoded_payment_18','encoded_payment_19','encoded_payment_20',
#                                    'encoded_payment_21','encoded_payment_22',
#                                    'encoded_payment_23','encoded_payment_24']

In [ ]:
# Before Filling Missing Values
#training_df[columns_with_missing_values].describe()

In [ ]:
# After Filling Missing Value
# training_df[columns_with_missing_values].describe()

In [ ]:
# training_df.info()
# No cells with missing Value

In [ ]:
# training_df.describe()

#### 2. Dropping unwanted Columns

In [ ]:
# drop_columns = ['customer_id','firstname','lastname']
# training_df = training_df.drop(columns=drop_columns)

In [14]:
def drop_columns(df):
    drop_columns = ['customer_id','firstname','lastname']
    df = df.drop(columns=drop_columns)
    return df

In [ ]:
# training_df = drop_columns(training_df)

3. Standardization/Normalization:

In [15]:
label = ['primary_close_flag', 'final_close_flag']

In [ ]:
    # for column in df.columns:
    #     if column not in label and pd.api.types.is_numeric_dtype(df[column]):
    #         data[column] = np.log1p(df[column])

    #scaler = StandardScaler()

In [16]:
def normalization(df):
    min_max_scaler = MinMaxScaler()
    #x_scaled = min_max_scaler.fit_transform(x)
    #df = pd.DataFrame(x_scaled)
    label = ['primary_close_flag', 'final_close_flag']

    for column in df.columns:
        if column not in label and pd.api.types.is_numeric_dtype(df[column]):
            df[column] = min_max_scaler.fit_transform(df[[column]])
    return df

In [ ]:
# training_df = normalization(training_df)

4. Convert Categorical Values as Numerals. (Encoding):

In [17]:
def categorical_to_numerals(df):
    columns_to_encode = [
      'is_zero_loans_within_5_days',
      'is_zero_loans_within_5_to_30_days',
      'is_zero_loans_within_30_to_60_days',
      'is_zero_loans_within_60_to_90_days',
      'is_zero_loans_over_90_days',
      'is_zero_utilization',
      'is_zero_over_limit_count',
      'is_zero_max_over_limit_count'
      ]

    encoder = LabelEncoder()

    for column in columns_to_encode:
        df[column] = encoder.fit_transform(df[column])
    #print('column : ',df.columns)

    return df

In [ ]:
# training_df = categorical_to_numerals(training_df)

In [18]:
def preprocessing(df):
    print("Starting Step 1 : fill_missing_value")
    df = fill_missing_value(df)
    print("Step 1 : fill_missing_value Completed")
    print("Starting Step 2 : dropping Columns")
    df = drop_columns(df)
    print("Step 2 : dropping Columns Completed")
    print("Starting Step 3 : normalization")
    df = normalization(df)
    print("Step 3 : normalization Completed")
    print("Starting Step 4 : categorical_to_numerals")
    df = categorical_to_numerals(df)
    print("Step 4 : categorical_to_numerals Completed")
    return df

### Doing Preprocessing

In [46]:
training_df = preprocessing(training_df)

Starting Step 1 : fill_missing_value
Step 1 : fill_missing_value Completed
Starting Step 2 : dropping Columns
Step 2 : dropping Columns Completed
Starting Step 3 : normalization
Step 3 : normalization Completed
Starting Step 4 : categorical_to_numerals
Step 4 : categorical_to_numerals Completed


In [20]:
testing_df = preprocessing(testing_df)

Starting Step 1 : fill_missing_value
Step 1 : fill_missing_value Completed
Starting Step 2 : dropping Columns
Step 2 : dropping Columns Completed
Starting Step 3 : normalization
Step 3 : normalization Completed
Starting Step 4 : categorical_to_numerals
Step 4 : categorical_to_numerals Completed


In [ ]:
# training_df.describe()

In [ ]:
# hist_plt = training_df.hist(figsize = (50,40),bins = 50)

In [ ]:
# training_df.describe()

In [ ]:
# flags= ['primary_close_flag','final_close_flag']
# training_df[flags].head(100)

## Exploratory Data Analysis
We See Correlation matrix, and other visualisation techniques to visuale the data to get understanding of the data.

In [ ]:
#TO_DO
# Correlation Matrix

In [ ]:
sum(training_df['primary_close_flag'] != training_df['final_close_flag'])/len(training_df)
# Checking Propotion of 2 Output Labels which different in primary and Final Flags

0.16817870483990768

In [21]:
def split_data():
    #Splitting the data for Training
    x = training_df.drop(columns=label)
    y = training_df[label]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=30)
    return x,y,x_train, x_test, y_train, y_test

In [22]:
print("splitting data")
x,y,x_train, x_test, y_train, y_test = split_data()

splitting data


Feature Selection


Correlation Factors


Variance ThresHolds

# Medelling and Training


## Generalised Function for Training and testing

Classifier Function :
--
Performance Metrices:
--
* accuracy,
* precision,
* recall, and
* F1-score

ROC_AUC Curve
--

In [23]:
def fit_model(classifier):
    class_model = classifier
    return class_model.fit(x_train,y_train)

In [24]:
def predict(class_model,data_set):
    return class_model.predict(data_set)

In [25]:
def print_accuracy(pred):
    # for classification we use accuracy and F1 score
    acc_score = accuracy_score(y_test,pred)
    print("Accuracy Score : ",acc_score)
    f_score = f1_score(y_test,pred,average='weighted')
    print("F1 Score : ",f_score)
    print("Multilabel Confusion Matrix:")
    print(multilabel_confusion_matrix(y_test, pred))
    print("Classification Report:")
    print(classification_report(y_test, pred))

In [26]:
def predict_score(classifier_name):
    #Calculate Score of Classifier on test data
    score = classifier_name.score(x_test, y_test)
    print("Score on Test Data: ",score)

In [27]:
def train_model(classifier):
    print("splitting data")
    x,y,x_train, x_test, y_train, y_test = split_data()
    print("fitting model")
    print("y_train")
    print(y_train.info())
    classifier_name = fit_model (classifier)
    # local()[var_name]
    print("Predicting x_test")
    predictions = predict(classifier_name,x_test)
    print("Printing accuracies")
    print_accuracy(predictions)
    predict_score(classifier_name)

    return classifier_name




```
def print_accuracy(pred,classifier_name):
    # for classification we use accuracy and F1 score
    acc_score = accuracy_score(y_test,pred)
    print("Accuracy Score : ",acc_score)
    f_score = f1_score(y_test,pred,average='weighted')
    print("F1 Score : ",f_score)
    print("Multilabel Confusion Matrix:")
    print(multilabel_confusion_matrix(y_test, pred))
    print("Classification Report:")
    print(classification_report(y_test, pred))
    #Calculate Score of Classifier on test data
    score = classifier_name.score(x_test, y_test)
    print("Score on Test Data: ",score)
```





```
def split_data():
    #Splitting the data for Training
    x = training_df.drop(columns=label)
    y = training_df[label]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=30)
    return x,y,x_train, x_test, y_train, y_test
```



```
def train_model(classifier):
    def split_data():
        #Splitting the data for Training
        x = training_df.drop(columns=label)
        y = training_df[label]
        x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=30)
        return x,y,x_train, x_test, y_train, y_test
    def fit_model(classifier):
        class_model = classifier
        return class_model.fit(x_train,y_train)
    def predict(class_model,data_set):
        return class_model.predict(data_set)
    def print_accuracy(pred):
        # for classification we use accuracy and F1 score
        acc_score = accuracy_score(y_test,pred)
        print("Accuracy Score : ",acc_score)
        f_score = f1_score(y_test,pred,average='weighted')
        print("F1 Score : ",f_score)
        print("Multilabel Confusion Matrix:")
        print(multilabel_confusion_matrix(y_test, pred))
        print("Classification Report:")
        print(classification_report(y_test, pred))
        #Calculate Score of Classifier on test data
        score = classifier_name.score(x_test, y_test)
        print("Score on Test Data: ",score)
    print("splitting data")
    x,y,x_train, x_test, y_train, y_test = split_data()
    print("fitting model")
    print("y_train")
    print(y_train.info())
    classifier_name = fit_model (classifier)
    # local()[var_name]
    print("Predicting x_test")
    predictions = predict(classifier_name,x_test)
    print("Printing accuracies")
    <!-- print_accuracy(predictions) -->
    print_accuracy(predictions,classifier_name)

    try:
        # Get the predicted probabilities or decision scores
        if hasattr(classifier_name, "predict_proba"):
            y_score = classifier_name.predict_proba(x_test)[:, 1]
        elif hasattr(classifier_name, "decision_function"):
            y_score = classifier_name.decision_function(x_test)
        else:
            raise RuntimeError("The classifier does not have a predict_proba or decision_function method.")

        # Calculate the false positive rates and true positive rates
        fpr, tpr, _ = roc_curve(y_test, y_score)
        # Calculate the AUC
        roc_auc = auc(fpr, tpr)

        # Plot the ROC curve
        plt.plot(fpr, tpr, label='ROC curve (area = %0.2f)' % roc_auc)
        plt.plot([0, 1], [0, 1], 'k--')
        plt.xlabel('False Positive Rate')
        plt.ylabel('True Positive Rate')
        plt.title('ROC Curve')
        plt.legend(loc="lower right")
        plt.show()
    except RuntimeError or TypeError as e:
        print(e)


    return classifier_name

```



Models Used:
--
*   Decision Trees (no.1)
*   Random Forests (no.2) -- More accurate then Decision Tree.
*   KNN Classifier

Then it is fine tuned for better result.

### 1. Decision Tree Classifier
Built manually.

In [ ]:
# Reindexing data before splitting data for Training.
#train_df = training_df.reindex(np.random.permutation(training_df.index))
# Not Req since, train_test_split contains random_state, which randomises the data before splitting

In [ ]:
# train_df = training_df so, training_df is not disturbed.
# train_df = training_df

In [ ]:
#Splitting the data for Training
X = training_df.drop(columns=label)
y = training_df[label]
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=30)
#test data contains 30% for testing, and random state = 30

In [ ]:
# DT_model = make_pipeline(MinMaxScaler(), DecisionTreeClassifier())
start_time = time.time()
DT_model= DecisionTreeClassifier()
DT_model.fit(x_train,y_train)
print('Time Take :  %s minutes' % ((time.time() - start_time)/60))

Time Take :  0.1996306856473287 minutes


In [ ]:
#Evaluating the model
# We use the predict() on the model to predict the output
start_time = time.time()
pred=DT_model.predict(x_test)
print('Time Take :  %s minutes' % ((time.time() - start_time)/60))

Time Take :  0.002860466639200846 minutes


In [ ]:
# for classification we use accuracy and F1 score
acc_score = accuracy_score(y_test,pred)
print("Accuracy Score : ",acc_score)

# accuracy = accuracy_score(y_test, predictions)
# print("Accuracy:", accuracy)
# print()

f_score = f1_score(y_test,pred,average='weighted')
print("F1 Score : ",f_score)

print("Multilabel Confusion Matrix:")
print(multilabel_confusion_matrix(y_test, pred))
print(" ")

print("Classification Report:")
print(classification_report(y_test, pred))

#Calculate Score of Classifier on test data
score = DT_model.score(x_test, y_test)
print("Score on Test Data: ",score)


Accuracy Score :  0.99983132682536
F1 Score :  0.9997123049378375
Multilabel Confusion Matrix:
[[[289234     23]
  [    29  42717]]

 [[268722      6]
  [     3  63272]]]
 
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     42746
           1       1.00      1.00      1.00     63275

   micro avg       1.00      1.00      1.00    106021
   macro avg       1.00      1.00      1.00    106021
weighted avg       1.00      1.00      1.00    106021
 samples avg       0.24      0.24      0.24    106021

Score on Test Data:  0.99983132682536


### 2. RandomForestClassifier

In [ ]:
RandomForestClassifier = train_model(RandomForestClassifier())

splitting data
fitting model
y_train
<class 'pandas.core.frame.DataFrame'>
Int64Index: 774671 entries, 314016 to 431909
Data columns (total 2 columns):
 #   Column              Non-Null Count   Dtype
---  ------              --------------   -----
 0   primary_close_flag  774671 non-null  int64
 1   final_close_flag    774671 non-null  int64
dtypes: int64(2)
memory usage: 17.7 MB
None
Predicting x_test
Printing accuracies
Accuracy Score :  0.99989457926585
F1 Score :  0.999830187299605
Multilabel Confusion Matrix:
[[[289251      6]
  [    26  42720]]

 [[268725      3]
  [     1  63274]]]
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     42746
           1       1.00      1.00      1.00     63275

   micro avg       1.00      1.00      1.00    106021
   macro avg       1.00      1.00      1.00    106021
weighted avg       1.00      1.00      1.00    106021
 samples avg       0.24      0.24      0.24    106021



### KNN Classifier ??
from sklearn.neighbors import KNeighborsClassifier
KNNClassifier = train_model(KNeighborsClassifier())

Time Taken : ???

This model is computationally too expensive, so, its better to take previous models and fine tune the models for better results.

In [28]:
# from sklearn.neighbors import KNeighborsClassifier
# KNNClassifier = train_model(KNeighborsClassifier())

### Classifier:

In [ ]:
params = {"n_neighbors": 5, "weights": "distance", "n_jobs": -1}
classifier = KNeighborsClassifier(**params)

In [ ]:
classifier

KNeighborsClassifier(n_jobs=-1, weights='distance')

In [ ]:
# classifier = KNeighborsClassifier()
print("fitting model")
print("y_train")
print(y_train.info())
classifier_name = fit_model (classifier)
# local()[var_name]
print("Predicting x_test")
predictions = predict(classifier_name,x_test)
print("Printing accuracies")
print_accuracy(predictions)
predict_score(classifier_name)

fitting model
y_train
<class 'pandas.core.frame.DataFrame'>
Int64Index: 774671 entries, 314016 to 431909
Data columns (total 2 columns):
 #   Column              Non-Null Count   Dtype
---  ------              --------------   -----
 0   primary_close_flag  774671 non-null  int64
 1   final_close_flag    774671 non-null  int64
dtypes: int64(2)
memory usage: 17.7 MB
None
Predicting x_test
Printing accuracies
Accuracy Score :  0.9328951846820661
F1 Score :  0.8846797280157356
Multilabel Confusion Matrix:
[[[283468   5789]
  [  4173  38573]]

 [[261773   6955]
  [  7641  55634]]]
Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.90      0.89     42746
           1       0.89      0.88      0.88     63275

   micro avg       0.88      0.89      0.88    106021
   macro avg       0.88      0.89      0.88    106021
weighted avg       0.88      0.89      0.88    106021
 samples avg       0.21      0.21      0.21    106021

Score on Tes

## Predicting O/P on testing_df for submission.

### Generalised Model

In [ ]:
def test_predict(model_obj,model_name):
    if type(model_name) != str:
        return false
    print("Predicting")
    predictions_sub = model_obj.predict(testing_df)
    print("DataFrame")
    df_submissions = pd.DataFrame({
        'primary_close_flag': predictions_sub[:, 0],
        'final_close_flag': predictions_sub[:, 1]
    })
    print("To_CSV")
    df_submissions.to_csv('/content/drive/MyDrive/Colab Notebooks/submit_'+model_name+'.csv', index=False)
    print("completed")

### Predicting for various Models & Submission
Calling test_predict function to predict and preparing csv for submission.

In [ ]:
# test_predict(LogisticRegressionClassifier,'LogisticRegressionClassifier')

In [ ]:
# test_predict(RandomForestClassifier,'RandomForestClassifier')

In [ ]:
# test_predict(LogisticRegressionClassifier,'LogisticRegressionClassifier')

In [ ]:
# test_predict(SVMClassifier_OvO,'SVMClassifier_OvO')

In [ ]:
# test_predict(SVMClassifier_OvR,'SVMClassifier_OvR')

In [ ]:
#MultinomialNB
# test_predict(MultinomialNB,'MultinomialNB')

In [ ]:
#GaussianNB
# test_predict(GaussianNB,'GaussianNB')

### Fine Tuning Models and Validating it with Test Dataset
We only use testing_df, for final testing and validation, only after training and finetuning the model with Train data set, which was split in ratio of 70:30 for Training and Testing, according to requirement, this is so as to avoid any form of Bias in model training which might arise, due to viewing or using provided Testing Dataset.

In [ ]:
#TO_DO
#Fine Tuning model

Fine Tuning Using RandomizedSearchCV for RandomForestClassifier

In [34]:
param_grid = {
    'n_estimators': [50, 100],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10]
}
start = time.time()
# Create a base model
rf = RandomForestClassifier()

# Instantiate the randomized search model
random_search_rf = RandomizedSearchCV(estimator=rf, param_distributions=param_grid, n_iter=100, cv=3, random_state=42,n_jobs=-1)

# Fit the randomized search to the data
random_search_rf.fit(x_train, y_train)
print('Time Take :  %s minutes' % ((time.time() - start)/60))

# Get the best parameters
best_params_rf = random_search_rf.best_params_
best_estimator_rf = random_search_rf.best_estimator_
best_score_rf = random_search_rf.best_score_

Time Take :  121.49984045823415 minutes


In [35]:
print('best_params_rf :',best_params_rf)
print('best_estimator_rf :',best_estimator_rf)
print('best_score_rf :',best_score_rf)

best_params_rf : {'n_estimators': 100, 'min_samples_split': 2, 'max_depth': None}
best_estimator_rf : RandomForestClassifier()
best_score_rf : 0.9998541316482376


Fine Tuning for DescisionTree

In [32]:
# Define the parameter grid
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10,15,20]
}
start_time = time.time()
# Create a base model
dt = DecisionTreeClassifier()

# Instantiate the randomized search model
random_search = RandomizedSearchCV(estimator=dt, param_distributions=param_grid, n_iter=100, cv=5,verbose=2, random_state=42,n_jobs=-1)
#First Try for cv = 3, but less computation
#later Try for cv = 5, but more computation
# Fit the randomized search to the data
random_search.fit(x_train, y_train)

# Get the best parameters
best_params_1 = random_search.best_params_
best_estimator_1 = random_search.best_estimator_
best_score_1 = random_search.best_score_

print('Time Take :  %s minutes' % ((time.time() - start_time)/60))


Fitting 5 folds for each of 24 candidates, totalling 120 fits
Time Take :  16.21891754468282 minutes


In [33]:
print('best_params_1 :',best_params_1)
print('best_estimator_1 :',best_estimator_1)
print('best_score_1 :',best_score_1)

best_params_1 : {'min_samples_split': 10, 'max_depth': None, 'criterion': 'entropy'}
best_estimator_1 : DecisionTreeClassifier(criterion='entropy', min_samples_split=10)
best_score_1 : 0.9998386411696641


In [31]:
!nproc --all

2


Fine Tuning for KNN

In [ ]:
# Define the parameter grid
param_grid = {
    'n_neighbors': list(range(1, 31)),
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']
}
start_time = time.time()
# Create a base model
knn = KNeighborsClassifier()

# Instantiate the randomized search model
random_search = RandomizedSearchCV(estimator=knn, param_distributions=param_grid, n_iter=100, cv=3, random_state=42)

# Fit the randomized search to the data
random_search.fit(x_train, y_train)
print('Time Take :  %s minutes' % ((time.time() - start_time)/60))
# Get the best parameters
best_params_2 = random_search.best_params_

### Fine Tuning Using Variance Method:


In [ ]:
#

## Testing the final model against Original Unused Test Dataset and Getting the score

From Fine Tuning, Best model is DecisionTreeClassifier,with params of criterion='entropy' and min_samples_split=10, where score is high, and computation is relatively low (Time), when compared with RandomForest which has similar score but Too high Computation.

In [38]:
x_train_final = training_df.drop(columns=label)
y_train_final = training_df[label]

In [53]:
#for Reference:
x_test_final = testing_df #Testing data without label used for predicting
y_test_final = submission_df # Test Data Set labels

In [39]:
final_model = DecisionTreeClassifier(criterion='entropy', min_samples_split=10)

In [40]:
start = time.time()
final_model.fit(x_train_final, y_train_final)
print('Time Take :  %s minutes' % ((time.time() - start)/60))

Time Take :  0.2692578514417013 minutes


In [41]:
print("Predicting")
start = time.time()
predictions_sub = final_model.predict(testing_df)
print('Time Take :  %s minutes' % ((time.time() - start)/60))
print("DataFrame")
dt_f_submissions = pd.DataFrame({
    'primary_close_flag': predictions_sub[:, 0],
    'final_close_flag': predictions_sub[:, 1]
})

Predicting
Time Take :  0.01828855276107788 minutes
DataFrame


### Creating Submit csv with Predicted Values.

In [42]:
print("To_CSV")
dt_f_submissions.to_csv('/content/drive/MyDrive/Colab Notebooks/submitDT_fine.csv', index=False)
print("completed")

To_CSV
completed


In [ ]:
test_predict(classifier_name,'KNNClassifier')

In [ ]:
print("Predicting")
predictions_sub = DT_model.predict(testing_df)
print("DataFrame")
df_submissions = pd.DataFrame({
    'primary_close_flag': predictions_sub[:, 0],
    'final_close_flag': predictions_sub[:, 1]
})


Predicting
DataFrame


In [ ]:
print("To_CSV")
df_submissions.to_csv('/content/submitDT.csv', index=False)
print("completed")

To_CSV
completed
